[FLINK-19503][state] Add StateChangelog API

rkhachatryan opened the pull request:
## What is the purpose of the change

Add State Changelog API and basic (in-memory) implementation as proposed in FLIP-158.

## Verifying this change

Added `StateChangelogClientTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

curcur commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/changelog/StateChangelogClientTest.java:
```diff
@@ -0,0 +1,158 @@
+            }
+
+            @Override
+            public StateChangeFormat getContext() {
```
"StateChangeFormat" is missing?

rkhachatryan replied:
This was a leftover after splitting the changes into two PRs, sorry about that.

curcur commented:
The StateChangeLog is put under flink-state-backends, 
under <module>flink-statebackend-changelog</module>

But this is where/the name I planned to put the WrappedStateBackend. What do you think?

rkhachatryan commented:
1. > But this is where/the name I planned to put the WrappedStateBackend.

In my opinion, both backend and changelog should reside in the same module, at least for now.

2. I fixed the `StateChangeFormat` issue

curcur commented:
> 1. > But this is where/the name I planned to put the WrappedStateBackend.
> 
> In my opinion, both backend and changelog should reside in the same module, at least for now.

I have no objection to putting them together.

> 
> 1. I fixed the `StateChangeFormat` issue

Thanks for fixing this!

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateChange.java:
```diff
@@ -0,0 +1,48 @@
+
+/** Change of state of a keyed operator. Used for generic incremental checkpoints. */
+@Internal
+public class StateChange {
```
This would suggest that there is `NonKeyedStateChange` somewhere. But currently there is none, and probably won't be. So it would be confusing to me.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateChangelogHandleStreamImpl.java:
```diff
@@ -0,0 +1,144 @@
+
+    private final KeyGroupRange keyGroupRange;
+    /** NOTE: order is important as it reflects the order of changes. */
+    private final List<Tuple2<StreamStateHandle, Long>> handlesAndOffsets;
```
> What is the second field, offset to the file?

Yes, the 2nd field in `handlesAndOffsets` is offset (to file or other handle).

> Would it be more clear to make it a separate class, with each field a specific name and java doc/comment?

To me, replacing `Tuple2<...>` with `HandleAndOffset` class isn't much better; and introducing a new abstraction like `StatePointer` isn't necessarily more clear.
Using `long` and `Tuple2` in `StateHandle` implementations is common, so I don't think we should change it.

> The structure might also change if we want to use something else (DB) instead of a blob DFS?

In that case, a new `StateChangelogHandle` implementation should be added.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateChangelogHandleStreamImpl.java:
```diff
@@ -0,0 +1,144 @@
+        return 0;
+    }
+
+    private static SharedStateRegistryKey getKey(StreamStateHandle stateHandle) {
```
> Or multiple backends bundles state changes altogether in a file

Exactly. For a single checkpoint, all backends of a single TM will put their changes into the same file. Reference count for that file will be `number of backends` initially.

If we instead use backend.uuid as a key then refCount would be 1 and file would be deleted as soon as one of the backends dereferences it.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/SequenceNumber.java:
```diff
@@ -0,0 +1,83 @@
+
+/**
+ * A logical timestamp to draw a boundary between the materialized and non-materialized changes.
+ * Maintained by the state backend but implementations may choose to move its generation to {@link
```
We can't put it into the state changelog because it is not aware of state materialization.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/StateChangelogWriter.java:
```diff
@@ -0,0 +1,77 @@
+    SequenceNumber lastAppendedSqn();
+
+    /** Appends the provided data to this log. No persistency guarantees. */
+    void append(int keyGroup, byte[] value);
```
Yes.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/StateChangelogWriter.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @param from inclusive
+     */
+    CompletableFuture<Handle> persist(SequenceNumber from) throws IOException;
```
Yes.

pnowojski commented on flink-state-backends/pom.xml:
```diff
@@ -39,5 +39,6 @@ under the License.
 	<modules>
 		<module>flink-statebackend-rocksdb</module>
 		<module>flink-statebackend-heap-spillable</module>
+		<module>flink-statebackend-changelog</module>
```
I think the interfaces definitions should be in the `flink-runtime`, probably along the in memory implementation. We know there are other usecases where we might want to use it, furthermore there will be different implementations of the changelog. 

Maybe file based implementation could be placed anywhere for the time being, but I would need to think about this structure a bit more.

rkhachatryan replied:
I'd say there **might** be use case where we could use. They are even close to design.
And it and **might** have different implementations.

I have concerns regarding both, but I don't mind putting the API part into the `flink-runtime`.

pnowojski commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/StateChangelogClient.java:
```diff
@@ -0,0 +1,38 @@
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.StateChangelogHandle;
+
+/**
+ * {@link StateChangelogWriter} factory. Scoped to a single entity (e.g. a SubTask or
+ * OperatorCoordinator).
+ */
+@Internal
+public interface StateChangelogClient<Handle extends StateChangelogHandle<?>>
```
Why is factory class named client? I had no idea what `StateChangelogClient` is doing looking at the name (I was more expecting it to be an internal class used by the changelog to perform the writes), and it's a bit funny that javadoc makes it easy to understand in the first sentence by calling it a `factory` :)

rkhachatryan replied:
Thanks for pointing out. This is a leftover from the previous design which assumed a dedicated component.
I'll rename it to `StateChangelogWriterFactory`.

pnowojski reviewed (commented):
I think we should wrap the loading of the concrete classes of the changelog as Plugins. It would both make it easier to provide a different implementations and would shield the system from unwanted dependencies. 

It should be fairly simple to do:
- have a factory class (we already do)
- load it using `PluginManager#load` 

All in all it's an easy thing to do initially (take a look at it's being used in the metric reporters for example `ReporterSetup#getAllReporterFactories`), but there is always a lot of resistance and a lot of problems to add it later on.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/StateChangelogWriter.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @param from inclusive
+     */
+    CompletableFuture<Handle> persist(SequenceNumber from) throws IOException;
```
It's from the gitven `from` until the last change already added.
I'll update the javadoc.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/changelog/StateChangelogWriter.java:
```diff
@@ -0,0 +1,77 @@
+    /**
+     * Get {@link SequenceNumber} of the last element added by {@link #append(int, byte[]) append}.
+     */
+    SequenceNumber lastAppendedSqn();
```
We've already agreed on this signature before but I can change it if you'd like.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogWriterFactoryLoader.java:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.Iterator;
+
+/** A thin wrapper around {@link PluginManager} to load {@link StateChangelogWriterFactory}. */
+@Internal
+public class StateChangelogWriterFactoryLoader {
+    private final PluginManager pluginManager;
+
+    public StateChangelogWriterFactoryLoader(PluginManager pluginManager) {
+        this.pluginManager = pluginManager;
+    }
+
+    @SuppressWarnings({"rawtypes"})
+    public Iterator<StateChangelogWriterFactory> load() {
+        return pluginManager.load(StateChangelogWriterFactory.class);
+    }
+}
```
What's the motivation/rationale behind this class?

rkhachatryan replied:
It should suggest users to rely on SPI/plugins instead of direct instantiation.
SPI part was missing which I guess confused you, sorry for that.

pnowojski replied:
optional nit: in that case I would add a sentence that changelog factories should be loaded using plugin mechanism instead of directly.

pnowojski reviewed (commented):
Thanks for the update @rkhachatryan . I think the code looks good % one last question.

Also maybe the package should be `org.apache.flink.runtime.state.changelog` for the API classes? But this is a small nit - it would become a valid/important question once we would make those classes `@Experimental` or `@PublicEvolving`.

rkhachatryan commented:
Thanks for reviewing @pnowojski .
I moved StateHandle classes to `org.apache.flink.runtime.state.changelog`. Other classes were already there.
I squashed the commits to remove confusing state with different packages.

rkhachatryan commented:
Test failure unrelated (FLINK-21277), merging.
